src/daemon/db_util.rs

```diff
@@ -0,0 +1,343 @@
+use crate::db::db_engine::{db_root, open_proxy_db};
+use crate::db::rolling::RollingDB;
+use crate::shim::{clock::ChainEpoch, version::NetworkVersion};
+use crate::utils::db::car_stream::CarStream;
+use crate::utils::{retry, RetryArgs};
+use anyhow::{bail, Context};
+use dialoguer::theme::ColorfulTheme;
+use futures::TryStreamExt;
+use positioned_io::RandomAccessFile;
+use std::ffi::OsStr;
+use std::fs;
+use std::{
+    path::{Path, PathBuf},
+    sync::Arc,
+    time,
+    time::Duration,
+};
+use tokio::io::AsyncWriteExt;
+use tracing::{error, info, warn};
+use url::Url;
+use walkdir::WalkDir;
+
+/// This function tries to open the forest database directory as [`ManyCar<Arc<RollingDB>>`], it
+/// 1. loads `parity-db`
+/// 2. loads all existing CAR files in `.forest.car.zst` format
+/// 3. asks to fetch the latest snapshot when it's required to run forest.
+/// 4. imports the snapshot(from either CLI options or step 3) and stores it to the database in `.forest.car.zst` format
+
+pub async fn open_forest_car_union_db(
+    config: &mut Config,
+    opts: &CliOpts,
+) -> anyhow::Result<(Arc<ManyCar<Arc<RollingDB>>>, Option<Tipset>)> {
+    let mut heaviest_tipset: Option<Tipset> = None;
+    let chain_data_path = chain_path(config);
+    let db_root_dir = db_root(&chain_data_path);
+    let forest_car_db_dir = db_root_dir.join("car_db");
+    if !forest_car_db_dir.is_dir() {
+        fs::create_dir_all(&forest_car_db_dir)?;
+    }
+
+    let mut store = ManyCar::new(Arc::new(open_proxy_db(
+        db_root_dir.clone(),
+        config.db_config().clone(),
+    )?));
+
+    let mut has_errors = false;
+    // Load existing CAR DB(s)
+    for file in WalkDir::new(&forest_car_db_dir)
+        .into_iter()
+        .filter_map(|entry| {
+            if let Ok(entry) = entry {
+                if let Some(filename) = entry.file_name().to_str() {
+                    if filename.ends_with(FOREST_CAR_FILE_EXTENSION) {
+                        return Some(entry.into_path());
+                    }
+                }
+            }
+            None
+        })
+    {
+        match ForestCar::new(RandomAccessFile::open(&file)?) {
+            Ok(car) => {
+                store.read_only(car.into());
+                info!("Loaded car DB at {}", file.display());
+            }
+            Err(err) => {
+                has_errors = true;
+                error!("Error loading car DB at {}: {err}", file.display())
+            }
+        };
+    }
+
+    // Let it check all DB files altogether instead of failing 1 by 1.
+    if has_errors {
+        bail!("Error loading DB file(s).");
+    }
+
+    // TODO: use `--consume-snapshot` CLI option once it's implemented
+    let mut consume_snapshot_file = false;
+    if config.client.snapshot_path.is_none() {
+        let epoch = {
+            if store.read_only_len() == 0 {
+                0
+            } else if let Ok(Some(ts)) = Tipset::load_heaviest(&store, &store) {
```

Comment:
This looks wrong. Can you explain this code to me?

Reply:
The idea is that when there are no CAR files loaded(manually deleted or whatever), the union DB should be treated as corrupted. Otherwise, we load the heaviest tipset from the store as normal

Reply:
Alternatively, I can inspect all tipsets back to genesis but that takes longer, is there a better way of achieving this?

```rust
async fn verify_tipsets_integrity(
    store: &impl Blockstore,
    from: Tipset,
    config: &Config,
) -> anyhow::Result<bool> {
    let start = Utc::now();
    info!("Verifying the database integrity...");
    let mut current = from;
    while current.epoch() > 0 {
        if let Some(parent) = Tipset::load(store, current.parents())? {
            current = parent;
        } else {
            break;
        }
    }

    let result = if current.epoch() > 0 {
        false
    } else {
        let genesis_header = read_genesis_header(
            config.client.genesis_file.as_ref(),
            config.chain.genesis_bytes(),
            store,
        )
        .await?;
        if let Some(genesis_ts) = Tipset::load(
            store,
            &TipsetKeys::new(FrozenCids::from_iter([*genesis_header.cid()])),
        )? {
            current == genesis_ts
        } else {
            false
        }
    };

    info!(
        "Done verifying the database integrity, took {}s",
        (Utc::now() - start).num_seconds()
    );

    Ok(result)
}
```